kubernetes/metrics.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// GetProcessedMetrics extracts metrics from k8s using GetRawMetrics and then populates them with labels and
+// float metrics that are converted from raw metrics' string values
+func GetProcessedMetrics(configPath string, namespace string, pods *PodMetricsList) error {
+	if err := GetRawMetrics(configPath, namespace, pods); err != nil {
+		return err
+	}
+
+	if err := GetLabels(configPath, namespace, pods); err != nil {
+		return err
+	}
+
+	ProcessMetrics(&pods)
+
+	return nil
+}
+
+// GetAverageMetrics extract metrics from k8s using GetRawMetrics and then transforms them to
+// new structure with labels and average CPU and memory per pod
+func GetAverageMetrics(configPath string, namespace string) ([]PodAverageMetrics, error) {
+	var pods PodMetricsList
+
+	if err := GetRawMetrics(configPath, namespace, &pods); err != nil {
+		return nil, err
+	}
+
+	if err := GetLabels(configPath, namespace, &pods); err != nil {
+		return nil, err
+	}
+
+	return CalculateAverageMetrics(&pods)
+}
+
+// GetLabels populates pod metrics list with labels
+func GetLabels(configPath string, namespace string, pods *PodMetricsList) error {
+	clientset, err := K8sClient.Get(configPath)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(pods.Items); i++ {
+		if len(pods.Items[i].Metadata.Labels) == 0 {
+			logging.Info("----getting labels for %v", pods.Items[i].Metadata.Name)
+			pod, err := clientset.CoreV1().Pods(namespace).Get(pods.Items[i].Metadata.Name, metav1.GetOptions{})
+			logging.Info("----got pod data: %v", pod)
+			if err != nil {
+				return err
+			}
+			pods.Items[i].Metadata.Labels = pod.Labels
+		}
+	}
+
+	return nil
+}
+
+// PodAverageMetrics provides average CPU and memor as long as some pod's metadata
+type PodAverageMetrics struct {
+	Name   string
+	Labels map[string]string
+	CPU    float64
+	Memory float64
+}
+
+func CalculateAverageMetrics(pods *PodMetricsList) ([]PodAverageMetrics, error) {
+	var res = make([]PodAverageMetrics, len(pods.Items))
+	for i := 0; i < len(pods.Items); i++ {
+		res[i].Name = pods.Items[i].Metadata.Name
+		res[i].Labels = pods.Items[i].Metadata.Labels
+		var sumCPU, sumMem float64
+		for j := 0; j < len(pods.Items[i].Containers); j++ {
+			cpu, err := ConvertCPU(pods.Items[i].Containers[j].Usage.CPU)
+			if err == nil {
+				sumCPU += cpu
+			} else {
+				logging.Error("Conversion of CPU for %v failed - %v", pods.Items[i], err)
+				return nil, err
+			}
+			mem, err := ConvertMemory(pods.Items[i].Containers[j].Usage.Memory)
+			if err == nil {
+				sumMem += mem
+			} else {
+				logging.Error("Conversion of Memory for %v failed - %v", pods.Items[i], err)
+				return nil, err
+			}
+		}
+		res[i].CPU = sumCPU / float64(len(pods.Items[i].Containers))
+		res[i].Memory = sumMem / float64(len(pods.Items[i].Containers))
+	}
+	return res, nil
+}
+
+// ProcessMetrics converts string values for CPU and memory to float ones and stores them into dedicated fields
+func ProcessMetrics(stract interface{}) {
```

Comment:
stract or struct?

Reply:
it's name of parameter and it cannot be reserved keyword (e.g. struct). Basically I wanted to show that it's intended to be struct internally that is passed as interface. But might be it's confusing.

Reply:
I'll rename it

Reply:
Renamed